cpp/include/cudf/detail/aggregation.hpp

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+
+namespace cudf {
+namespace experimental {
+
+/**
+ * @brief Base class for specifying the desired aggregation in an
+ * `aggregation_request`.
+ *
+ * Other kinds of aggregations may derive from this class to encapsulate
+ * additional information needed to compute the aggregation.
+ */
+class aggregation {
+ public:
+  /**
+   * @brief Possible aggregation operations
+   */
+  enum Kind { SUM, MIN, MAX, COUNT, MEAN, MEDIAN, QUANTILE };
```

Comment:
Why is this class in a `detail` file? Seems to me that the type of the aggregations is really part of the main user API, so should be in the non-detail header. Evidence for this is that it's not in the detail namespace.

Reply:
It's an opaque(ish) type. There's no reason for users to know about the internals of the class, so it's only forward declared in the public headers. The detail header contains the definition.